OpenTrons/opentrons-api
# refactor(robot-server): Store protocols in an in-memory SQL database

# Overview

This is a bit of incremental work towards #8468.

Formerly, `robot-server`'s `ProtocolStore` is backed by a regular Python `dict`. With this PR, it's backed by an in-memory, non-persistent SQLite database. This is a very roundabout way of accomplishing exactly the same thing. But it gets us pretty close to persisting protocol uploads across robot reboots.

# Testing

The work so far is covered well by existing unit tests.

If you want to test manually, you'll have to do it on a local dev server. Until https://github.com/Opentrons/buildroot/pull/151 is merged, this will crash `robot-server` when running on a real OT-2.

1. Upload some protocols with several files.
2. Inspect the responses from the upload, and from `GET /protocols` and `GET /protocols/{id}`. Make sure they all seem correct: no weird datetimes, no unexpected `null`s, and so on.

# Risk assessment

*For now,* fairly low, because this work so far is scoped to a single unit that's covered well by tests. But things will get spicy when we start actually persisting things.

# Next steps and big open questions

## Making this actually persistent

* Currently, we store protocol files in [a temporary directory](https://github.com/Opentrons/opentrons/blob/0375ca35881a869376b6ebacdbcdd776f1fcc5b7/robot-server/robot_server/protocols/dependencies.py#L58). We'll need to change this to make it somewhere persistent. The simplest way of doing this is a 1-line change.
* Currently, we store `ProtocolResource`s in an in-memory database. We'll need to change this to store them in an on-disk file. 

## Taking `ProtocolSource` out of the database

Edit: Moved to #9858.

## Persisting protocols' analyses?

Edit: After discussion, we on CPX don't think we should do this. #8468 updated with details.

## Moving to SQLAlchemy ORM?

As written, this only uses SQLAlchemy Core, avoiding the ORM, because that's what we said we would do in the outcome of https://github.com/Opentrons/opentrons/issues/8496. But implementing it this way turns out to involve a lot of type-unsafe mapping between SQLAlchemy columns and our own internal dataclasses. Maybe we should use the ORM after all. We'd still have to map between SQLAlchemy types and our own types, but at least it would be a little more type-safe. [Theoretically](https://docs.sqlalchemy.org/en/14/orm/extensions/mypy.html), anyway. 

Edit: After discussion, "maybe." We're leaving this alone for now.
